### PR TITLE
31 auth from GitHub and session part 1

### LINKED ISSUE #31: auth : from github and session (part 1)
# Description 
we need to be able to identify users, and give them a space in app to interact with the platform
we will use github authentication because all our users will be contributing to code projects during their mentoring

### why not email/password : 
because we will have many things covered for us:
- email verification flow
- forgot password  flow
- upload of a profile picture
- the users will not have to fill a form, makes it easier to create an account

# acceptance criteria
- provide an endpoint to launch the authentication flow
- user will agree to allow ioprodz to authenticate him and get his profile
- user is generated a tokens (jwt), set to cookies on the users device
- cookies should be 
  - secured scoped to ioprodz.com 
  - https only
- the two tokens are :
  - access token (TTL 5min) containing : sub <user uuid>, email, scopes <array of scopes> (with other jwt values)
  - refresh token (TTL 24h) containing : sub <user uuid>, tokenId <token uuid>

# tech spec
## routes 
`GET /auth/github`
will start the auth process with github
`POST /auth/github/callback`
will get the confirmation of github
`POST /auth/refresh`
will create a session token, and send now auth tokens

## controllers
- 1 :  receive the auth request and pass it to githubAuthStrategy.authenticate()
- 2 : process the github callback
  - if user profile (by provider id) does not exist , create him one, if not update his profile with new data
  - create a session for him, (should be by device) (yes we store the User-Agent in the session)
  - send him an access token and a refresh (session) token
  
docs: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#web-application-flow

- 3 : process refresh token
  - decode token, update session.lastusedAt in db (if fail : send 401)
  - send him an access token

## business
```ts
interface AuthStrategy {
   authenticate()
   handleAuthentication()
}
```

```ts
class GithubAuthStrategy implements AuthStrategy {
   // implementation
}
```

```ts
class JwtAuthStrategy implements AuthStrategy {
   // implementation
}
```
## storage

- we keep 3 set of information :
  - identity will contain : id, email, provider <name of authority ex: github>, 
  - profile : id, identityId, name, dob, <we will explore what github gives>, avatarUrl
  - sessions: id, ttl: <int seconds of validity , startedAt<date>, lastusedAt <date>

### PR DESCRIPTION
**Pull request by badido39**

### CODE REVIEW

File: hoggar/src/app/config.ts [hunk 1]
```diff
@@ -4,4 +4,5 @@ dotenv.config();
 
 export const port = parseInt(process.env.PORT || "8000", 10);
 export const hostname = process.env.HOSTNAME || "localhost";
-export const dburl = process.env.DATABASE_URL || "file:data/db.sqlite";
+export const dburl = process.env.DATABASE_URL || "file:data/dev.db";
+export const CLIENT_ID = process.env.CLIENT_ID || "59bacf304331ecf0c890";
```
Comment (osminosm) [inline hoggar/src/app/config.ts]:
```suggestion
export const dburl = process.env.DATABASE_URL || "file:data/db.sqlite";
export const CLIENT_ID = process.env.CLIENT_ID;
```

**Review by osminosm — CHANGES_REQUESTED** [review]

**Review by nazimboudeffa — APPROVED** [review]

File: hoggar/prisma/migrations/20221105155546_initial/migration.sql [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    "bio" TEXT NOT NULL DEFAULT '',
+    "githubUser" TEXT NOT NULL DEFAULT '',
+    "avatarUrl" TEXT NOT NULL DEFAULT '',
+    "websiteUrl" TEXT NOT NULL DEFAULT '',
+    "location" TEXT NOT NULL DEFAULT '',
+    "company" TEXT NOT NULL DEFAULT '',
+    CONSTRAINT "Profile_identityId_fkey" FOREIGN KEY ("identityId") REFERENCES "Identity" ("id") ON DELETE CASCADE ON UPDATE CASCADE
+);
+
+-- CreateIndex
```
Comment (osminosm) [inline hoggar/prisma/migrations/20221105155546_initial/migration.sql]: Still need index on identity (provider, providerId)

File: hoggar/src/auth/routes.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      const ghAccessToken = await getAccessToken(code as string);
+      const { data } = await getUserData(ghAccessToken);
+      const identity = await ensureIdentityByProvider(
+        "github",
+        data.id,
+        data.email
+      );
+      await updateProfileByIdentityId(identity.id, githubProfileAdapter(data));
+      const tokens = await createSessionForIdentity(identity.id, userAgent);
+      res.status(201).json(tokens);
```
Comment (osminosm) [inline hoggar/src/auth/routes.ts]: need to set tokens to secure cookies

Comment (badido39) [reply hoggar/src/auth/routes.ts]: you have to add app.use(cookieParser());

Comment (osminosm) [reply hoggar/src/auth/routes.ts]: yes for when we need to parse

Comment (badido39) [reply hoggar/src/auth/routes.ts]: in our case the app is declared in src/app/index.ts , here we can add all uses

Comment (badido39) [reply hoggar/src/auth/routes.ts]: and then we can implement in response 
res.cookie("access_token", token, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production"})

File: hoggar/src/auth/business.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      identityId,
+    },
+  });
+
+  const access_token = jwt.sign({ sub: identityId }, "secret", {
+    expiresIn: 300, // 5 minutes
+  });
+  const refresh_token = jwt.sign(
+    { sub: identityId, sid: session.id },
+    "secret",
```
Comment (osminosm) [inline hoggar/src/auth/business.ts]: secret from config AUTH_JWT_SECRET

File: hoggar/prisma/migrations/20221105155546_initial/migration.sql (hunk 2 above)
Comment (badido39) [reply hoggar/prisma/migrations/20221105155546_initial/migration.sql]: i see the code i think there is no need to index betwen (provider , providerId)

**Review by osminosm — APPROVED** [review]

**Pull request merged**